server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -234,23 +239,16 @@ class AsyncAction {
             this.task = task;
             this.request = request;
             this.listener = listener;
-            if (request.concreteNodes() == null) {
-                resolveRequest(request, clusterService.state());
-                assert request.concreteNodes() != null;
-            }
-            this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
-            this.concreteNodes = request.concreteNodes();
-
-            if (request.getIncludeDiscoveryNodes() == false) {
-                // As we transfer the ownership of discovery nodes to route the request to into the AsyncAction class, we
-                // remove the list of DiscoveryNodes from the request. This reduces the payload of the request and improves
-                // the number of concrete nodes in the memory.
-                request.setConcreteNodes(null);
-            }
+            this.concreteNodes = resolveRequest(request, clusterService.state());
+            this.responses = new AtomicReferenceArray<>(this.concreteNodes.length);
         }
 
         void start() {
             final DiscoveryNode[] nodes = this.concreteNodes;
+            // As we transfer the ownership of discovery nodes to route the request to into the AsyncAction class,
+            // we remove the list of DiscoveryNodes from the request. This reduces the payload of the request and improves
+            // the number of concrete nodes in the memory.
+            request.setConcreteNodes(null);
```

Comment:
- Why is this required - given that `resolveRequest` is no longer setting the `concreteNodes`. 
- Can the concreteNodes be removed from request always ? There might be some actions which might require setting it in request for stable nodeId's

Reply:
> Why is this required - given that resolveRequest is no longer setting the concreteNodes.

We need this because there are few callers of TransportNodesAction which pass a specific list of nodes to which transport request has to be sent.
E.g. - In AsyncShardFetch we identify the nodes to fetch from based on the cache. 
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L179

>Can the concreteNodes be removed from request always ? There might be some actions which might require setting it in request for stable nodeId's

Yes. Currently except for TransportNodesReloadSecureSettingsAction no other transport node actions is using this. As part of this PR, we are also removing dependency of TransportNodesReloadSecureSettingsAction from  concreteNodes object.

Reply:
Got it, so it is set either in the constructor (or) as part of resolveRequest.

I think we can put back the boolean flag `IncludeDiscoveryNodes` and reset to null conditionally. This is to avoid breaking any external plugins that might extend the BaseNodeRequest.